### ISSUE TITLE
GUI : eBMC Server Power Operations

### ISSUE THREAD

**Issue by derick-montague** [ibm-openbmc/dev#973]
## Stakeholders
SME: @geissonator
PHYP SME: David Stanton, Travis Pizel, Pete Heyrman
IBM i SME: Kris Whitney
Designers: @derick-montague, @ParishrutB, @priyanka-pillai97 

### Summary
The oBMC Server power operations works for all Linux systems regardless of vendor. TThe existing page has a setting called Boot Override and an option called One time boot. Both of these are only available in OPAL mode. These functions **will not work on a Rainier system that are PowerVM based.**

### Functionality
See feature discovery document listed below for screen shots and details functionality information.

- Power On
- Orderly Shutdown
- Warm Reboot
- Cold Reboot
- ~~Keylock~~ 
    - Determined not needed in #1840 
- IPL Mode
- Default Partition Mode
- ~~Timed Power On~~
  - Not functionality in ASMI today
  - Handled either through HMC or Server OS

### InVision Prototype
Below are all the different design iterations for this story. As of 8/9/21, the latest design is not tested of listed. 

The following tasks are required:
- [x] Update design with feedback from design and development meeting
- [x] Add the latest design to Current concepts
- [x] Test with internal users

Final Design: https://ibm.invisionapp.com/share/TCO1HH637ZF

---
**Current concepts for Operating mode**
Concept 1: https://ibm.invisionapp.com/d/main#/console/15366764/320038298/preview
Concept 3 7/28: https://ibm.invisionapp.com/d/main#/console/15366764/320053619/preview
Concept 4 7/28: https://ibm.invisionapp.com/d/main#/console/15366764/320051916/preview
Concept 5 7/29: https://ibm.invisionapp.com/d/main#/console/15366764/320054742/preview
Concept for Power Restore Policy page: https://ibm.invisionapp.com/d/main#/console/15366764/320051925/preview
Concept for Overview page: https://ibm.invisionapp.com/d/main#/console/15366764/320051922/preview

**Old prototypes**
HMC-managed: https://ibm.invisionapp.com/share/8ENYRVXAPFD#/319219002_07-Server-On-Off
Non-HMC managed: https://ibm.invisionapp.com/share/8ENYRVXAPFD#/screens/319212965
Non-HMC managed IBM i: https://ibm.invisionapp.com/share/8ENYRVXAPFD#/319216076_10-Server-On-Off

### References
- [Feature Discovery](https://ibm.box.com/s/nff263oru77wkrmqbjqs3gx8v1qdjgo3)
- [Slack channel conversation about Manual and Normal boot settings](https://ibm-systems-power.slack.com/archives/C0Q6TQP5Z/p1588339060248600)
- [Slack channel conversation about BIOS attribute availability](https://ibm-systems-power.slack.com/archives/C0Q6TQP5Z/p1610688096016700)

### UX Flow / Interaction Requirements
- [ ] Feature breakdown

### cURL Commands and Redfish Response
- [ ] Add Redfish response
- [ ] Add cURL commands

### Notes
- Check with engineering to see if we can also implement #1795 as part of this story. If so, we can close #1795.
- This will be a page that is behind the IBM environmental flag
- There are some [new state being added to Redfish](https://github.com/ibm-openbmc/dev/issues/2619#issue-716098074) and it will be good to determine their purpose and if they impact our current UI.
- The Server power operating mode has 4 properties that require a specific setting. The Manual or automatic setting is a BIOS attribute called `pvm_system_operating_mode` 
    - The other properties will be set using [3 different REST End points defined in this GitHub comment](https://github.com/ibm-openbmc/dev/issues/973#issuecomment-886238533) within this story. _If in ZenHub, this link needs to be opened in a new window._

#### Browser Tests (Chrome, Firefox, Safari (Mac), Edge (Windows))
- [ ] Tested responsive layout on supported viewport sizes
- [ ] Verified functionality works as expected, e.g. success and failure use cases
- [ ] Resolved any JavaScript errors thrown to the console

#### Accessibility Tests
- [ ] Verified markup is semantic
- [ ] No reported issues in IBM Accessibility Checker (Free open source tool) https://www.ibm.com/able/toolkit/tools/

#### Test Hooks
- [ ] Verify need for `data-test-id` hooks with test team

![image (5).png](https://images.zenhubusercontent.com/5c969c26c2488c52df50b7eb/dc0dd25d-cf4a-42f7-a5db-cfbc1a346b6d)

**Comment by rfrandse** [comment]
refresh

**Comment by ryanarnell** [comment]
SME notes- https://ibm.ent.box.com/notes/443193157118

**Comment by derick-montague** [comment]
One suggestion we need to discuss with SME's is keeping the navigation naming, `Server Power Operations` rather than changing to `Server On/Off Operations`

**Comment by ryanarnell** [comment]
Keylock is mostly known as Server Operating Mode in ASMI and in pour design here

**Comment by derick-montague** [comment]
@ryanarnell - We may need to look back into this based on this Slack conversation: 
https://ibm-systems-power.slack.com/archives/C0Q6TQP5Z/p1585215425100800

**Comment by derick-montague** [comment]
I think we should reopen this story based on this [conversation in slack about System Operating mode](https://ibm-systems-power.slack.com/archives/GBVDZRCQK/p1588352403238100)

**Comment by nicoleconser** [comment]
Per [7/28 Slack conversation](https://ibm-systems-power.slack.com/archives/C0Q6TQP5Z/p1595954021097800), double-check with Travis Pizel on whether the “default” option can be removed from the **Default Partition Environment** list (only shows up on certain box types).

**Comment by nicoleconser** [comment]
note: According to Pete Heyrman, if we decide to keep the **Default Partition Environment** section on this page, then we **do not** need to keep the **Hardware Page Table Size** function in the GUI. Current expectation is that we will keep the Default Partition Environment on this screen, meaning we can remove HPT Size.

**Comment by derick-montague** [comment]
@priyanka-pillai97 let's discuss how BIOS settings will impact the design of these pages.

**Comment by derick-montague** [comment]
We have Timed Power on removed in the description. I noticed we are working on timed power on in #1467 and #1468. Does that change that assumption?

**Comment by derick-montague** [comment]
- Verified non-stand alone systems proposed UI

**Comment by derick-montague** [comment]
[Discussion today in slack ](https://ibm-systems-power.slack.com/archives/GBVDZRCQK/p1607439382046400?thread_ts=1607435058.041200&cid=GBVDZRCQK)regarding components that are not part of the prototypes since they were discovered following the prototype. The two that may need to be added are:
1. Enable / Disable Virtual TPM
1. Enabled / Disable Secure Storage / Key lock
1. Previous boot progress indicator should be a consideration as well

This has also been captured in the [feature discovery document](https://ibm.box.com/s/nff263oru77wkrmqbjqs3gx8v1qdjgo3).

**Comment by Srisuk** [comment]
Till now we have the data for 2 BIOS attributes which are:
1.pvm_system_power_off_policy
2.pvm_default_os_type

3 attributes whose information are still not available:
1.pvm_power_on_st
2.pvm_rpa_boot_mode
3.pvm_system_operating_mode
Discussion of slack is mentioned below:
https://ibm-systems-power.slack.com/archives/C0Q6TQP5Z/p1612282080283600?thread_ts=1610688096.016700&cid=C0Q6TQP5Z

**Comment by Srisuk** [comment]
Resolved by https://gerrit.openbmc-project.xyz/c/openbmc/webui-vue/+/40389

**Comment by derick-montague** [comment]
[Slack Conversations regarding concerns about Manual mode settings](https://ibm-systems-power.slack.com/archives/C0Q6TQP5Z/p1611245845012100?thread_ts=1610688096.016700&cid=C0Q6TQP5Z)

> Manual mode includes 3 different Redfish API’s. 2 of them are already present and a third is under discussion with Redfish. The issue I see is that these settings are also individual customer settings in the GUI. I’m not sure how the GUI will be able to remember the customer settings vs. the manual mode override of those settings. It may be we just have to display a warning to the user that enabling Manual mode will override those settings.

**Comment by Srisuk** [comment]
Currently the missing keys from BE are:
1. pvm_system_operating_mode
2. pvm_rpa_boot_mode
3. pvm_power_on_st
4. pvm_firmware_boot_side
5. pvm_os_ipl_type

**Comment by derick-montague** [comment]
**Blocked**: Need to have BIOS attributes implemented on the backend.

**Comment by Srisuk** [comment]
Currently the BIOS attributes which arent available through API:

**Comment by derick-montague** [comment]
Scrum of scrums today:
> Waiting on PHYP to implement. Maury planning to discuss in Dean's meeting tonight.

**Comment by Srisuk** [comment]
Backend code is merged as part of https://gerrit.openbmc-project.xyz/c/openbmc/pldm/+/41715, waiting to test the feature in machine. https://ibm-systems-power.slack.com/archives/GBVDZRCQK/p1619150950076100?thread_ts=1618416613.075500&cid=GBVDZRCQK

**Comment by gtmills** [comment]
"Manual mode" will be 3 different APIs 2 of which are there today. The 3rd is tracked by #1068

**Comment by Srisuk** [comment]
@gtmills currently the implementation for `Server Power Policy` is based upon the value from backend BIOS key `pvm_system_power_off_policy`, the API used are:

1.To GET BIOS key value: `/redfish/v1/Systems/system/Bios`
2.To GET BIOS key attribute values: `/redfish/v1/Registries/BiosAttributeRegistry/BiosAttributeRegistry`

The implementation is dynamic for these values, it's difficult to map and render the data if it is coming from multiple keys.

**Comment by derick-montague** [comment]
@Srisuk - What @gtmills has shared is updates to the Redfish API, so for manual/normal mode, it has been decided to use these and the 3rd API instead of the BIOS attribute. I am working with Gunnar to figure out our next steps. I am concerned that this will result in the need for design changes.

**Comment by derick-montague** [comment]
```
curl -k -X PATCH -d '{"PowerRestorePolicy":"AlwaysOff"}' https://${bmc}/redfish/v1/Systems/system
curl -k -v -X PATCH https://${BMC_IP}/redfish/v1/Systems/system -d '{"Boot":{"AutomaticRetryConfig": "Disabled"}}'
curl -k -H "X-Auth-Token: $token" -X PATCH -d '{ "Boot" :{"StopBootOnFault":"AnyFault"}}' https://${bmc}/redfish/v1/Systems/system
```

**Comment by derick-montague** [comment]
Determined in the last Design exchange meeting that the eBMC engineering team would like to move away from the server operating mode nomenclature established on FSP-based systems in ASMI. Those systems have two server operating modes, `Normal` and `Manual`. The manual mode performed the following in ASMI:

**Comment by derick-montague** [comment]
Moved into code review in order to be able to test the changes needed to support System operating mode with users.

**Comment by derick-montague** [comment]
Discussed on July 21st and July 27th in the Design Exchange meeting. 

[Documented in the design exchange wiki](https://syswiki02.rchland.ibm.com/openbmc/openbmc-design-reviews)

**Comment by dixsie** [comment]
@Srisuk pvm_system_operating_mode is available in the driver

**Comment by dixsie** [comment]
@Srisuk `StopBootOnFault` contains two values `Never` and `AnyFault`. For the checkbox I assume `never` = `false` and `AnyFault` =  `true`. 

Add support for redfish StopBootOnFault: https://gerrit.openbmc-project.xyz/c/openbmc/bmcweb/+/44175

So if checkbox is selected patch:
```curl -k -H "X-Auth-Token: $token" -X PATCH -d
'{ "Boot" :{"StopBootOnFault":"AnyFault"}}'
https://${bmc}/redfish/v1/Systems/system
```
If checkbox is unselected then patch: 
```curl -k -H "X-Auth-Token: $token" -X PATCH -d
'{ "Boot" :{"StopBootOnFault":"Never"}}'
https://${bmc}/redfish/v1/Systems/system
```

**Comment by dixsie** [comment]
@Srisuk `AutomaticRetryConfig` has 3 values, Disabled, RetryAlways, and
RetryAttempts. OpenBMC only supports `Disabled and RetryAttempts`.

if checkbox selected: 
```
curl -k -v  -X PATCH https://${bmc}/redfish/v1/Systems/system -d \
'{"Boot":{"AutomaticRetryConfig": "RetryAttempts"}}'
```

if checkbox unselected:
```
curl -k -v  -X PATCH https://${bmc}/redfish/v1/Systems/system -d \
'{"Boot":{"AutomaticRetryConfig": "Disabled"}}'```

**Comment by Srisuk** [comment]
Resolved by https://gerrit.openbmc-project.xyz/c/openbmc/webui-vue/+/47013

**Comment by Srisuk** [comment]
Resolved by https://github.ibm.com/openbmc/webui-vue/pull/28

**Comment by mzipse** [comment]
Work is complete.
